pex/pep_376.py

```diff
@@ -0,0 +1,430 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import base64
+import csv
+import errno
+import fileinput
+import hashlib
+import json
+import os
+import shutil
+import sys
+from contextlib import closing
+from fileinput import FileInput
+
+from pex import dist_metadata
+from pex.common import is_python_script, safe_mkdir
+from pex.compatibility import MODE_READ_UNIVERSAL_NEWLINES, PY2, get_stdout_bytes_buffer, urlparse
+from pex.enum import Enum
+from pex.orderedset import OrderedSet
+from pex.third_party.pkg_resources import Distribution, EntryPoint
+from pex.typing import TYPE_CHECKING, cast
+from pex.util import CacheHelper
+from pex.venv.virtualenv import Virtualenv
+
+if TYPE_CHECKING:
+    if PY2:
+        from hashlib import _hash as _Hash
+    else:
+        from hashlib import _Hash
+    from typing import Callable, Dict, Iterable, Iterator, Optional, Protocol, Tuple, Union
+
+    import attr  # vendor:skip
+
+    class CSVWriter(Protocol):
+        def writerow(self, row):
+            # type: (Iterable[Union[str, int]]) -> None
+            pass
+
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class Digest(object):
+    algorithm = attr.ib()  # type: str
+    encoded_hash = attr.ib()  # type: str
+
+    def new_hasher(self):
+        # type: () -> _Hash
+        return hashlib.new(self.algorithm)
+
+
+@attr.s(frozen=True)
+class Hash(object):
+    value = attr.ib()  # type: str
+
+    def __str__(self):
+        # type: () -> str
+        return self.value
+
+    def parse(self):
+        # type: () -> Digest
+        algorithm, encoded_hash = self.value.split("=", 1)
+        return Digest(algorithm=algorithm, encoded_hash=encoded_hash)
+
+
+@attr.s(frozen=True)
+class InstalledFile(object):
+    """The record of a single installed file from a PEP 376 RECORD file.
+
+    See: https://www.python.org/dev/peps/pep-0376/#record
+    """
+
+    path = attr.ib()  # type: str
+    hash = attr.ib(default=None)  # type: Optional[Hash]
+    size = attr.ib(default=None)  # type: Optional[int]
+
+
+class InstallationScheme(Enum["InstallationScheme.Value"]):
+    """Represents the Pip installation scheme used for installing a wheel.
+
+    For more about installation schemes, see:
+        https://docs.python.org/3/install/index.html#alternate-installation
```

Comment:
Is this the right link to put here?  The link refers to distutils but these schemes are pip flags, so that link is a bit confusing (to me at least) in this context. E.g., naively, I don't see --root or --target mentioned there but I do see --home ?

Reply:
That's the definition at any rate. Pip comes up with custom schemes but those are only defined in code.